### DISCUSSION TITLE
Custom renderpass in CallbackTrait?

### DISCUSSION THREAD

**Discussion by tao-hengyi** [emilk/egui#4583] [Q&A]
I have read the [custom3d_wgpu](https://github.com/emilk/egui/blob/master/crates/egui_demo_app/src/apps/custom3d_wgpu.rs) example. I now want to extend this example by adding features such as shadows that require multiple render passes. 

However, in the parameters of the paint method of `CallbackTrait`, there is only the main egui render pass.

```rust
fn paint<'a>(
    &'a self,
    info: PaintCallbackInfo,
    render_pass: &mut wgpu::RenderPass<'a>,
    callback_resources: &'a CallbackResources,
);
```

In this case, how can I implement multiple render passes?

**Comment by 9SMTM6** [comment]
I'm neither a webgpu nor egui expert. But can't you simply render this using the full wgpu in like another thread, and then render the final result in egui?

A decent bit more effort, but i dont see why that would not work, and it'd be arguably more portable.

**Comment by mkeeter** [comment]
I have gotten this working, and figured I'd post details (since this is the top search result for "egui custom render pass").

In my case, I needed to render images with a depth buffer attached.  The solution is basically what @9SMTM6 suggests: render to a texture, then draw that texture in `egui`.  However, you **don't** need to go off-thread; it's possible to do this all using `egui_wgpu::CallbackTrait`.

In `CallbackTrait::prepare`, you're given a `&mut wgpu::CommandEncoder`.  You can use this command encoder to start your own render pass, with whatever settings you want (e.g. depth culling).  Importantly, this render pass should use a texture (which you create!) as its color attachment.  `egui` will run this render pass before the render pass which draws the UI.

Then, in `CallbackTrait::paint`, you blit the resulting texture to the screen as a full-screen quad.

You'll need two (or more) pipelines: all of your fancy custom rendering gets done in your custom render pass with fancy pipelines, then the `CallbackTrait::paint` pipeline just draws the texture.

In my case, I needed to create both RGBA and depth textures, but your needs may vary.  The RGBA texture should be configured with `usage: TextureUsages::RENDER_ATTACHMENT | TextureUsages::TEXTURE_BINDING`, so it can be both a render target and used in drawing.
